### PR TITLE
Deprecate `squeeze` in GroupBy.

### LINKED ISSUE #2157: groupby should not squeeze out dimensions
#### Code Sample

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([1, 3, 6])),
    }
)
list(arr.groupby('x'))

[(1, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 1), 
(3, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 3), 
(6, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 6)]
```
#### Problem description

The dimension _x_ disappear. I have done some tests and it seems that this problem raise only with strictly ascending coordinates.
For example in this case it works correctly:

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([2, 1, 0])),
    }
)
list(arr.groupby('x'))

[(0, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 0), 
(1, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 1), 
(2, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 2)]
```

#### Expected Output

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([1, 3, 6])),
    }
)
list(arr.groupby('x'))

[(1, <xarray.DataArray (x: 1)>
  ar1ay([1.])
  Coordinates:
    * x        (x) int64 1), 
(3, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 3),
 (6, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 6)]
```

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-41-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.4
pandas: 0.22.0
numpy: 1.14.3
scipy: 1.1.0
netCDF4: 1.3.1
h5netcdf: None
h5py: 2.7.1
Nio: None
zarr: None
bottleneck: None
cyordereddict: None
dask: 0.17.4
distributed: 1.21.8
matplotlib: 2.2.2
cartopy: 0.16.0
seaborn: None
setuptools: 38.4.1
pip: 10.0.1
conda: None
pytest: 3.5.1
IPython: 6.2.1
sphinx: 1.7.4

</details>

### LINKED ISSUE #8263: Surprising `.groupby` behavior with float index
### What is your issue?

We raise an error on grouping without supplying dims, but not for float indexes — is this intentional or an oversight?

> This is without `flox` installed

```python

da =  xr.tutorial.open_dataset("air_temperature")['air']

da.drop_vars('lat').groupby('lat').sum()
```

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[8], line 1
----> 1 da.drop_vars('lat').groupby('lat').sum()
...
ValueError: cannot reduce over dimensions ['lat']. expected either '...' to reduce over all dimensions or one or more of ('time', 'lon').
```

But with a float index, we don't raise:

```python
da.groupby('lat').sum()
```

...returns the original array:

```
Out[15]:
<xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)>
array([[[296.29   , 296.79   , 297.1    , ..., 296.9    , 296.79   ,
         296.6    ],
        [295.9    , 296.19998, 296.79   , ..., 295.9    , 295.9    ,
         295.19998],
        [296.6    , 296.19998, 296.4    , ..., 295.4    , 295.1    ,
         294.69998],
...
```

And if we try this with a non-float index, we get the error again:

```python
da.groupby('time').sum()
```

```
ValueError: cannot reduce over dimensions ['time']. expected either '...' to reduce over all dimensions or one or more of ('lat', 'lon').
```

### LINKED ISSUE #8518: Groupby reduction fails when all groups are of size 1
### What happened?

When grouping by a dimension with no duplicate coordinates, reduction operations fail. Reductions don't fail if there are duplicate coordinates, or if an empty tuple is provided in the reduction operation.

### What did you expect to happen?

Reductions happening after a groupby operation which leads to each group being a singleton should not fail. By default it should mimic the same behaviour as when an empty tuple is passed to the reduction operation.

### Minimal Complete Verifiable Example

```Python
import numpy as np
import xarray as xr

a = xr.DataArray(range(3), coords={'a': [0, 1, 1]})
b = xr.DataArray(range(3), coords={'a': [0, 1, 2]})

# Works completely fine
a.groupby('a').sum()

# What is expected from the below:
b.groupby('a').sum(tuple())

# Raises value error
b.groupby('a').sum()
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

```Python
ValueError                                Traceback (most recent call last)
File /private/tmp/foo.py:11, in <module>
      8 a.groupby('a').sum()
     10 # Raises value error
---> 11 b.groupby('a').sum()
     13 # What is expected from the above:
     14 b.groupby('a').sum(tuple())

File ~/miniconda3/lib/python3.10/site-packages/xarray/core/_aggregations.py:6204, in DataArrayGroupByAggregations.sum(self, dim, skipna, min_count, keep_attrs, **kwargs)
   6194     return self._flox_reduce(
   6195         func="sum",
   6196         dim=dim,
   (...)
   6201         **kwargs,
   6202     )
   6203 else:
-> 6204     return self.reduce(
   6205         duck_array_ops.sum,
   6206         dim=dim,
   6207         skipna=skipna,
   6208         min_count=min_count,
   6209         keep_attrs=keep_attrs,
   6210         **kwargs,
   6211     )

File ~/miniconda3/lib/python3.10/site-packages/xarray/core/groupby.py:1478, in DataArrayGroupByBase.reduce(self, func, dim, axis, keep_attrs, keepdims, shortcut, **kwargs)
   1468 def reduce_array(ar: DataArray) -> DataArray:
   1469     return ar.reduce(
   1470         func=func,
   1471         dim=dim,
   (...)
   1475         **kwargs,
   1476     )
-> 1478 check_reduce_dims(dim, self.dims)
   1480 return self.map(reduce_array, shortcut=shortcut)

File ~/miniconda3/lib/python3.10/site-packages/xarray/core/groupby.py:69, in check_reduce_dims(reduce_dims, dimensions)
     67     reduce_dims = [reduce_dims]
     68 if any(dim not in dimensions for dim in reduce_dims):
---> 69     raise ValueError(
     70         f"cannot reduce over dimensions {reduce_dims!r}. expected either '...' "
     71         f"to reduce over all dimensions or one or more of {dimensions!r}."
     72     )

ValueError: cannot reduce over dimensions ['a']. expected either '...' to reduce over all dimensions or one or more of ().
```

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:03:09) [Clang 13.0.1 ]
python-bits: 64
OS: Darwin
OS-release: 23.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: ('en_GB', 'UTF-8')
libhdf5: 1.12.1
libnetcdf: None

xarray: 2023.6.0
pandas: 1.5.3
numpy: 1.23.5
scipy: 1.8.1
netCDF4: None
pydap: None
h5netcdf: None
h5py: 3.6.0
Nio: None
zarr: None
cftime: None
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: 1.3.5
dask: 2022.8.0
distributed: 2022.8.0
matplotlib: 3.5.2
cartopy: None
seaborn: None
numbagg: None
fsspec: 2023.4.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 67.8.0
pip: 21.3.1
conda: 4.13.0
pytest: None
mypy: None
IPython: 8.4.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] xref #2157, xref #1460 
- [x] Closes #8518, closes #8263
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -73,6 +74,21 @@ def check_reduce_dims(reduce_dims, dimensions):
...
 
 
+def _maybe_squeeze_indices(
+    indices, squeeze: bool | None, grouper: ResolvedGrouper, warn: bool
+):
+    if squeeze in [None, True] and grouper.can_squeeze:
+        if squeeze is None and warn:
+            emit_user_level_warning(
+                "The `squeeze` kwarg to GroupBy is being removed."
+                "Pass .groupby(..., squeeze=False) to silence this warning."
```
Comment (max-sixty) [inline xarray/core/groupby.py:84]: I think we should say that something is happening that won't be happening; it's not a case where the only change will be the warning going away, IIUC

```suggestion
                "Pass .groupby(..., squeeze=False) to disable squeezing, and silence this warning."
```

(or maybe "disable squeezing" needs more description, not sure)

**Review by max-sixty — APPROVED** [review]
+1! Thanks!

**Comment by dcherian** [comment]
This unfortunately ended up being a lot more invasive, to minimize warnings raised. 
Could use a second round of review.

File: xarray/core/groupby.py [hunk 2]
```diff
@@ -71,9 +71,27 @@ def check_reduce_dims(reduce_dims, dimensions):
...
+                f" Try passing .groupby(..., squeeze=False)"
             )
 
 
+def _maybe_squeeze_indices(
+    indices, squeeze: bool | None, grouper: ResolvedGrouper, warn: bool
+):
+    if squeeze in [None, True] and grouper.can_squeeze:
+        if isinstance(indices, slice):
+            if indices.stop - indices.start == 1:
```
Comment (headtr1ck) [inline xarray/core/groupby.py:83]: Is that save?
What about None or negative Start-Stop Values?

Comment (dcherian) [reply xarray/core/groupby.py:83]: Both are not possible today:
` group_indices: T_GroupIndices = [slice(i, i + 1) for i in range(size)]` at Line 455

and at Line 553:
```
group_indices: T_GroupIndices = [
            slice(i, j) for i, j in zip(sbins[:-1], sbins[1:])
]
```

Comment (dcherian) [reply xarray/core/groupby.py:83]: Ah no you're right, there's an edge case in TimeResamplerGrouper where `stop` is None, and we're resampling to the same frequency as the data so `grouper.can_squeeze` is True.

Comment (dcherian) [reply xarray/core/groupby.py:83]: Ah no it doesn't run, but i have no explicitly skipped squeezing for resampling. This is the current behaviour on `main`

**Comment by max-sixty** [comment]
Great work @dcherian !

**Pull request merged**